taichi/program/kernel.cpp

```diff
@@ -266,6 +266,22 @@ void Kernel::LaunchContextBuilder::set_arg_external_array(
   ctx_->set_device_allocation(arg_id, is_device_allocation);
 }
 
+void Kernel::LaunchContextBuilder::set_arg_external_array_w_shape(
+    int arg_id,
+    uint64 ptr,
+    uint64 elem_size,
```

Comment:
can we pass in elem_size * elem_count directly?

Reply:
I pass them separately for the reason that multiplication in python scope has noticeable overhead. 
The overhead would be accumulated to significance in the case of frequent kernel launch, so I made a slightly different interface. 

Ideally I should pass the "Ndarray" object directly to the C++ function in order to eliminate the extra three or four pybind invokes (`get_data_ptr``get_nelement``get_element_size` are just pybind calls). However, I encountered a pybind problem of reference passing from the python world is still a little verbose and painful. I'll leave it here if I can not fix this at noon.

Reply:
Solved by 1) pass in elem_size * elem_count directly.
 2) add set function for Ndarray to further reduce overhead.